### PR TITLE
argparse: Add use_argument_group=True

### LINKED ISSUE #3076: Support Trainer.add_argparse_args for ArgumentGroup
## 🚀 Feature
Support `Trainer.add_argparse_args` for `argparse._ArgumentGroup`

### Motivation

Adding all of the Trainers args to an existing parser clutters its help message with many arguments. Ideally, I would like to have these in a separate argument group.

### Pitch

Currently:
```python
parser = argparse.ArgumentParser()
parser.add_argument("--my_arg")
parser = pl.Trainer.add_argparse_args(parser)
parser.print_help()
```
```bash
usage: script.py [-h] [--my_arg MY_ARG] [--logger [LOGGER]] [--checkpoint_callback [CHECKPOINT_CALLBACK]] [--early_stop_callback [EARLY_STOP_CALLBACK]] [--default_root_dir DEFAULT_ROOT_DIR] [--gradient_clip_val GRADIENT_CLIP_VAL] (... and many more)

optional arguments:
  -h, --help            show this help message and exit
  --my_arg MY_ARG
  --logger [LOGGER]     autogenerated by pl.Trainer
  --checkpoint_callback [CHECKPOINT_CALLBACK]
                        autogenerated by pl.Trainer
  --early_stop_callback [EARLY_STOP_CALLBACK]
                        autogenerated by pl.Trainer
  --default_root_dir DEFAULT_ROOT_DIR
                        autogenerated by pl.Trainer
  --gradient_clip_val GRADIENT_CLIP_VAL
                        autogenerated by pl.Trainer
  (... and many more)

```

Desired:
```python
parser = argparse.ArgumentParser()
parser.add_argument("--my_arg")
pl_parser = parser.add_argument_group(title="pytorch-lightning Trainer arguments")
pl_parser = pl.Trainer.add_argparse_args(pl_parser)
parser.print_help()
```
```bash
usage: script.py [-h] [--my_arg MY_ARG] [--logger [LOGGER]] [--checkpoint_callback [CHECKPOINT_CALLBACK]] [--early_stop_callback [EARLY_STOP_CALLBACK]] [--default_root_dir DEFAULT_ROOT_DIR] [--gradient_clip_val GRADIENT_CLIP_VAL] (... and many more)

optional arguments:
  -h, --help            show this help message and exit
  --my_arg MY_ARG

pytorch-lightning Trainer arguments:
  --logger [LOGGER]     autogenerated by pl.Trainer
  --checkpoint_callback [CHECKPOINT_CALLBACK]
                        autogenerated by pl.Trainer
  --early_stop_callback [EARLY_STOP_CALLBACK]
                        autogenerated by pl.Trainer
  --default_root_dir DEFAULT_ROOT_DIR
                        autogenerated by pl.Trainer
  --gradient_clip_val GRADIENT_CLIP_VAL
                        autogenerated by pl.Trainer
  (... and many more)
```

### LINKED ISSUE #5875: Improve interface between Trainer and argparse
## 🚀 Feature

Current Trainer.add_argparse_args(parser_param) creates a new ArgumentParser object, passing the parser_param object as a parent to the newly created parser. Trainer.add_argparse_args() returns the new ArgumentParser object to its caller.

An improved interface would simply add all arguments to parser_param without creating a new object. To maintain backwards compatibility with almost all existing code, Trainer.add_argparse_args() may simply return the parser_param object.

### Motivation

The current interface makes impossible using advanced argparse features like grouping and subparsing. Due to the huge number of parameters in Trainer, isolating them in a group (instead of leaving them intermixed with everything else) is a common use case.

### Pitch

A small change in `pytorch_lightning.utilities.argparse_utils.add_argparse_args`, essentially removing the one line that creates a new ArgparseParser object:
```
parser = ArgumentParser(parents=[parent_parser], add_help=False,)
```

### Alternatives

The only alternatives I could think of are horrible: 

1. Clients subclassing Trainer and overriding add_argparse_args (breaking abstraction);
2. Clients calling Trainer.add_argparse_args with a clean parser and using protected members of ArgparseParser to copy the arguments onto a group or subparser of the actual parser (breaking abstraction in an even worse way).

### LINKED ISSUE #6086: (minor, ui) argparse stuff - add `add_argparse_args(*, inplace)` option for multiple subparsers?
### To Reproduce

See script and setup / version stuff.

```py
import argparse

import pytorch_lightning as pl

def main():
    parser = argparse.ArgumentParser()
    parser_main = parser.add_argument_group("main")
    parser_main.add_argument("--my_custom_arg_plz", type=str)
    parser_pl = pl.Trainer.add_argparse_args(
        parser.add_argument_group("pl.Trainer"),
    )
    # Doesn't have pl args.
    print("[ parser ]\n")
    parser.print_help()
    # Has pl.Trainer args, but, uh, didn't actually print any sections :(
    print("\n\n[ parser_pl ]\n")
    parser_pl.print_help()


assert __name__ == "__main__"
main()

"""
Setup: CPython 3.6.9, Ubuntu 18.04

    cd $(mktemp -d)
    python3 -m venv .
    ./bin/pip install -U pip wheel
    ./bin/pip install pytorch-lightning==1.2.0 torch==1.7.1

Output (abbreviated):

    $ ./bin/python ./example.py
    [ parser ]

    usage: example.py [-h] [--my_custom_arg_plz MY_CUSTOM_ARG_PLZ]

    optional arguments:
      -h, --help            show this help message and exit

    main:
      --my_custom_arg_plz MY_CUSTOM_ARG_PLZ

    [ parser_pl ]
    usage: example.py [-h] [--my_custom_arg_plz MY_CUSTOM_ARG_PLZ]
                      [--logger [LOGGER]]
                      [--checkpoint_callback [CHECKPOINT_CALLBACK]]
                      <... soooo many things...>

    optional arguments:
      -h, --help            show this help message and exit
      --my_custom_arg_plz MY_CUSTOM_ARG_PLZ
      --logger [LOGGER]     Logger (or iterable collection of loggers) for
                            experiment tracking.
      --checkpoint_callback [CHECKPOINT_CALLBACK]
      <... soooo many things...>
"""
```

### Expected behavior

Output would ideally look like this:

```
usage: example.py [-h] [--my_custom_arg_plz MY_CUSTOM_ARG_PLZ]
                  [--logger [LOGGER]]
                  [--checkpoint_callback [CHECKPOINT_CALLBACK]]
                  <... soooo many things...>

optional arguments:
  -h, --help            show this help message and exit

main:
      --my_custom_arg_plz MY_CUSTOM_ARG_PLZ

pl.Trainer:
  --logger [LOGGER]     Logger (or iterable collection of loggers) for
                        experiment tracking.
  --checkpoint_callback [CHECKPOINT_CALLBACK]
  <... soooo many things...>
```

### Environment

You can get the script and run it with:
```
wget https://raw.githubusercontent.com/PyTorchLightning/pytorch-lightning/master/tests/collect_env_details.py
# For security purposes, please check the contents of collect_env_details.py before running it.
python collect_env_details.py
```

 - PyTorch Version (e.g., 1.0): 1.7.1
 - OS (e.g., Linux): Linux, Ubuntu 18.04
 - How you installed PyTorch (`conda`, `pip`, source): pip
 - Build command you used (if compiling from source): N/A
 - Python version: CPython 3.6.9
 - CUDA/cuDNN version: Doesn't matter.
 - GPU models and configuration: Doesn't matter.
 - Any other relevant information: See repro script.

### Additional context

I maybe should've done more testing in #5911 :sweat_smile:

### PR DESCRIPTION
**Pull request by EricCousineau-TRI**
## What does this PR do?

Fixes #6086
Closes #5851
Fixes #3076
Fixes #5875

## Before submitting
- [/] Was this discussed/approved via a GitHub issue? Not really, but simple 'nuff?
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [x] Did you make sure to update the documentation with your changes? (if necessary)
- [x] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- (N/A) Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
Mayyyyybe. :hand_over_mouth:

### CODE REVIEW

File: pl_examples/domain_templates/generative_adversarial_net.py [hunk 1]
```diff
@@ -130,8 +130,11 @@ def __init__(
         self.example_input_array = torch.zeros(2, self.hparams.latent_dim)
 
     @staticmethod
-    def add_argparse_args(parent_parser: ArgumentParser):
-        parser = ArgumentParser(parents=[parent_parser], add_help=False)
+    def add_argparse_args(parent_parser: ArgumentParser, inplace=False):
+        if inplace:
+            parser = parent_parser
+        else:
+            parser = ArgumentParser(parents=[parent_parser], add_help=False)
```
Comment (Borda) [inline pl_examples/domain_templates/generative_adversarial_net.py:137]:
```suggestion
        parser = parent_parser if inplace else ArgumentParser(parents=[parent_parser], add_help=False)
```

File: pytorch_lightning/core/datamodule.py [hunk 2]
```diff
@@ -23,7 +23,10 @@
 
 from pytorch_lightning.core.hooks import CheckpointHooks, DataHooks
 from pytorch_lightning.utilities import parsing, rank_zero_only
-
+from pytorch_lightning.utilities.argparse import (
+    add_argparse_args,
+    from_argparse_args,
+)
```
Comment (Borda) [inline pytorch_lightning/core/datamodule.py:29]:
```suggestion
from pytorch_lightning.utilities.argparse import add_argparse_args, from_argparse_args
```

**Review by Borda — APPROVED** [review]
lgtm

File: pl_examples/domain_templates/generative_adversarial_net.py (hunk 1 above)
Comment (EricCousineau-TRI) [reply pl_examples/domain_templates/generative_adversarial_net.py]: Added! IMO less readable, but will defer!

File: pytorch_lightning/core/datamodule.py [hunk 3]
```diff
@@ -248,82 +248,12 @@ def setup(self, stage: Optional[str] = None):
...
-            args = cls.parse_argparser(args)
-        params = vars(args)
-
-        # we only want to pass in valid DataModule args, the rest may be user specific
-        valid_kwargs = inspect.signature(cls.__init__).parameters
-        datamodule_kwargs = dict((name, params[name]) for name in valid_kwargs if name in params)
-        datamodule_kwargs.update(**kwargs)
-
-        return cls(**datamodule_kwargs)
+        return from_argparse_args(cls, args, **kwargs)
```
Comment (carmocca) [inline pytorch_lightning/core/datamodule.py:256]: Keep the docstring

File: pytorch_lightning/core/datamodule.py [hunk 4]
```diff
@@ -248,82 +248,12 @@ def setup(self, stage: Optional[str] = None):
...
-                dest=arg,
-                default=arg_default,
-                type=use_type,
-                help=f"autogenerated by plb.{cls.__name__}",
-                **arg_kwargs,
-            )
-
-        return parser
+    def add_argparse_args(cls, parent_parser: ArgumentParser, **kwargs) -> ArgumentParser:
+        return add_argparse_args(cls, parent_parser, **kwargs)
```
Comment (carmocca) [inline pytorch_lightning/core/datamodule.py:252]: Keep the docstring

File: docs/source/common/trainer.rst [hunk 5]
```diff
@@ -130,7 +130,9 @@ So you can run it like so:
         parser = ArgumentParser()
         parser = Trainer.add_argparse_args(
             # group the Trainer arguments together
-            parser.add_argument_group(title="pl.Trainer args")
+            parser.add_argument_group(title="pl.Trainer args"),
+            # add it to the origianl parser
```
Comment (akihironitta) [inline docs/source/common/trainer.rst:134]: [nitpicking]
```suggestion
            # add it to the original parser
```

File: tests/trainer/test_trainer_cli.py [hunk 6]
```diff
@@ -71,11 +72,34 @@ def test_add_argparse_args_redefined(cli_args):
 def test_add_argparse_via_argument_group(cli_args):
     """Simple test ensuring that passing an argument group still works"""
     parser = ArgumentParser(add_help=False)
-    parser = Trainer.add_argparse_args(parser.add_argument_group(title="pl.Trainer args"))
+    parser = Trainer.add_argparse_args(
```
Comment (tchaton) [inline tests/trainer/test_trainer_cli.py:75]: Can we automate this in Trainer.add_argparse_args ? I don't except users to know about this.

**Review by Borda — COMMENTED** [review]
@EricCousineau-TRI mind check the last comments 🐰

File: tests/trainer/test_trainer_cli.py (hunk 6 above)
Comment (carmocca) [reply tests/trainer/test_trainer_cli.py:75]: I wouldn't make the choice to use `ArgumentGroup` for the user.

**Review by carmocca — APPROVED** [review]

File: tests/trainer/test_trainer_cli.py (hunk 6 above)
Comment (EricCousineau-TRI) [reply tests/trainer/test_trainer_cli.py:75]: Agreed that it should be up to the user.
(Personally, I would prefer it, but there may be weird edge cases if this is done by default at this point... mebbe good for future PR?)

File: pytorch_lightning/utilities/argparse.py [hunk 7]
```diff
@@ -135,29 +134,44 @@ def get_init_arguments_and_types(cls) -> List[Tuple[str, Tuple, Any]]:
...
+        >>> parser = argparse.ArgumentParser()
+        >>> parser = Trainer.add_argparse_args(parser.add_argument_group("pl.Trainer"), inplace=True)
+        >>> args = parser.parse_args([])
     """
-    parser = ArgumentParser(
-        parents=[parent_parser],
-        add_help=False,
-    )
+    if inplace:
+        parser = parent_parser
```
Comment (EricCousineau-TRI) [inline pytorch_lightning/utilities/argparse.py:169]: Nuts. CI fails here 'cause this logic is just wrong. Will fix.

Comment (EricCousineau-TRI) [reply pytorch_lightning/utilities/argparse.py:176]: Erg blech: b55521e5
Fixes it, but adds some nasty edge cases. Sorry :sweat_smile: 

Can a PL dev resolve this convo if they find it satisfactory? (or, well, marginally acceptable?)

File: pytorch_lightning/utilities/argparse.py [hunk 8]
```diff
@@ -135,29 +136,49 @@ def get_init_arguments_and_types(cls) -> List[Tuple[str, Tuple, Any]]:
...
     Examples:
+
+        # Option 1: Default usage.
         >>> import argparse
         >>> from pytorch_lightning import Trainer
         >>> parser = argparse.ArgumentParser()
         >>> parser = Trainer.add_argparse_args(parser)
         >>> args = parser.parse_args([])
+
+        # Option 2: Using add_argument_group(). Warning: Do not use the return value to parse args!
```
Comment (carmocca) [inline pytorch_lightning/utilities/argparse.py:170]: Add a why to the warning, "As the argument group is returned, not the parser"

File: pytorch_lightning/utilities/argparse.py [hunk 9]
```diff
@@ -135,29 +136,49 @@ def get_init_arguments_and_types(cls) -> List[Tuple[str, Tuple, Any]]:
     return name_type_default
 
 
-def add_argparse_args(cls, parent_parser: ArgumentParser) -> ArgumentParser:
+def add_argparse_args(
+    cls,
+    parent_parser: Union[ArgumentParser, ArgumentGroup],
+    *,
+    inplace=False,
+) -> Union[ArgumentParser, ArgumentGroup]:
```
Comment (carmocca) [inline pytorch_lightning/utilities/argparse.py:144]: Update the signatures of the functions that call this one, please.

File: pytorch_lightning/utilities/argparse.py (hunk 7 above)
Comment (carmocca) [reply pytorch_lightning/utilities/argparse.py:176]: Fine with me, left some more comments

File: tests/trainer/test_trainer_cli.py (hunk 6 above)
Comment (EricCousineau-TRI) [reply tests/trainer/test_trainer_cli.py]: Maybe I spoke too soon. After latest push, maybe it would be best to just make this the default?

Then all the extra stuff, warnings, etc. all disappear. The only warning is to not do `cls.add_argparse_args(parser.add_argument_group("..."))`, and instead just the normal way.

Thoughts @carmocca / @Borda? It would simplify user interaction TBH.

File: pytorch_lightning/utilities/argparse.py [hunk 10]
```diff
@@ -19,6 +19,8 @@
 
 from pytorch_lightning.utilities import parsing
 
+ArgumentGroup = type(ArgumentParser().add_argument_group("fake"))
```
[the remark this answers is not in the document]
Comment (EricCousineau-TRI) [reply pytorch_lightning/utilities/argparse.py:22]: Main concern is that it's private, so may not be the most stable. But yeah, I can change it.

File: tests/trainer/test_trainer_cli.py (hunk 6 above)
Comment (carmocca) [reply tests/trainer/test_trainer_cli.py]: Well that'd open the door for somebody to open another PR in a few months so he doesn't have to use `ArgumentGroup`. I would like to avoid that

If you can add a flag to disable `ArgumentGroup`, then that's okay

**Comment by tchaton** [comment]
Hey @EricCousineau-TRI,

You have some failing tests. Would you mind addressing them,  so we can merge the PR ?

Best,
T.C

File: pytorch_lightning/utilities/argparse.py (hunk 9 above)
Comment (EricCousineau-TRI) [reply pytorch_lightning/utilities/argparse.py]: Done. (No longer relevant, I think)

File: pytorch_lightning/utilities/argparse.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (EricCousineau-TRI) [reply pytorch_lightning/utilities/argparse.py]: Done.

File: tests/trainer/test_trainer_cli.py (hunk 6 above)
Comment (EricCousineau-TRI) [reply tests/trainer/test_trainer_cli.py]: Done. Added explicit `use_argument_group=True` arg. (avoiding double-neg)

**Comment by EricCousineau-TRI** [comment]
> You have some failing tests. Would you mind addressing them, so we can merge the PR ?

I think I fixed 'em! :crossed_fingers:

File: pytorch_lightning/utilities/argparse.py [hunk 11]
```diff
@@ -184,6 +221,8 @@ def add_argparse_args(cls, parent_parser: ArgumentParser) -> ArgumentParser:
         else:
             use_type = arg_types[0]
 
+        # TODO: Defer these to an optional static method, like
+        # `argparse_hacks()`?
```
Comment (carmocca) [inline pytorch_lightning/utilities/argparse.py:225]: I'd rather not put this TODO as it will be easily forgotten. Feel free to open a PR for it in the future!

**Review by carmocca — COMMENTED** [review]
I really like the update, a few more failing tests to go!

File: pytorch_lightning/core/datamodule.py [hunk 12]
```diff
@@ -248,58 +249,13 @@ def setup(self, stage: Optional[str] = None):
...
-                default=arg_default,
-                type=use_type,
-                help=f"autogenerated by plb.{cls.__name__}",
-                **arg_kwargs,
-            )
-
-        return parser
+    def add_argparse_args(cls, parent_parser: ArgumentParser, **kwargs) -> ArgumentParser:
+        """Extends existing argparse by default `LightningDataModule` attributes."""
+        return add_argparse_args(cls, parent_parser, **kwargs)
```
[the remark this answers is not in the document]
Comment (EricCousineau-TRI) [reply pytorch_lightning/core/datamodule.py:254]: Fixed this, but new error, still dunno root-cause:
```
$ python -m pytest -v pl_examples/test_examples.py::test_examples_cpu --exitfirst
...
any.py: error: unrecognized arguments: --batch_size 32
```

File: pytorch_lightning/utilities/argparse.py (hunk 11 above)
Comment (EricCousineau-TRI) [reply pytorch_lightning/utilities/argparse.py]: Done in b385eefb

File: pytorch_lightning/core/datamodule.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (EricCousineau-TRI) [reply pytorch_lightning/core/datamodule.py:254]: Done. 9c548ea47c1c22c570cfa319df44f58f87c55a5f

blech

File: pytorch_lightning/utilities/argparse.py [hunk 13]
```diff
@@ -135,37 +134,84 @@ def get_init_arguments_and_types(cls) -> List[Tuple[str, Tuple, Any]]:
...
     allowed_types = (str, int, float, bool)
 
+    # Get symbols from cls or init function.
+    for symbol in (cls, cls.__init__):
+        args_and_types = get_init_arguments_and_types(symbol)
+        args_and_types = [x for x in args_and_types if x[0] not in ignore_arg_names]
+        if len(args_and_types) > 0:
+            break
+
     args_help = parse_args_from_docstring(cls.__init__.__doc__ or cls.__doc__)
```
Comment (EricCousineau-TRI) [inline pytorch_lightning/utilities/argparse.py:212]: New CI failure:
```
$ python -m pytest -v tests/core/test_datamodules.py::test_dm_add_argparse_args
...
E       AttributeError: 'NoneType' object has no attribute 'split'
```

Comment (EricCousineau-TRI) [reply pytorch_lightning/utilities/argparse.py]: Done 79129533

**Comment by EricCousineau-TRI** [comment]
Passed (finally!)

File: pytorch_lightning/utilities/argparse.py [hunk 14]
```diff
@@ -135,37 +134,84 @@ def get_init_arguments_and_types(cls) -> List[Tuple[str, Tuple, Any]]:
...
+    assert isinstance(cls, type), repr(cls)
+    if cls.__module__.startswith("pytorch_lightning."):
+        # Abbreviate.
+        return f"pl.{cls.__name__}"
+    else:
+        # Fully qualified.
+        return f"{cls.__module__}.{cls.__qualname__}"
+
+
+def add_argparse_args(
```
Comment (awaelchli) [inline pytorch_lightning/utilities/argparse.py:147]: did you only move this or did you make changes too?

File: tests/utilities/test_argparse_utils.py [hunk 15]
```diff
@@ -48,3 +59,112 @@ def test_parse_args_from_docstring_empty():
...
+def test_negative_add_argparse_args():
+    with pytest.raises(RuntimeError, match="Please only pass an ArgumentParser instance."):
+        parser = ArgumentParser()
+        add_argparse_args(AddArgparseArgsExampleClass, parser.add_argument_group("bad workflow"))
+
+
+def test_add_argparse_args_no_argument_group():
+    """
+    Test similar to above, but with ``use_argument_group=False``
+    (old workflow).
```
Comment (awaelchli) [inline tests/utilities/test_argparse_utils.py:153]: could you describe it without referring to "above"?

**Review by awaelchli — APPROVED** [review]

File: pytorch_lightning/utilities/argparse.py (hunk 14 above)
Comment (EricCousineau-TRI) [reply pytorch_lightning/utilities/argparse.py:147]: Didn't mean to move (silly git diff!), but had inserted the `get_abbrev_qualified_cls_name` function above, and added `add_argument_group` option.

File: tests/utilities/test_argparse_utils.py (hunk 15 above)
Comment (EricCousineau-TRI) [reply tests/utilities/test_argparse_utils.py]: Done in 285f2ec0

**Comment by EricCousineau-TRI** [comment]
Confused by codecov failures:
- https://github.com/PyTorchLightning/pytorch-lightning/pull/6088/checks?check_run_id=2015464436
- https://github.com/PyTorchLightning/pytorch-lightning/pull/6088/checks?check_run_id=2015464216

Will try re-merging master.

**Comment by carmocca** [comment]
Sadly, codecov hasn't been working properly lately, feel free to ignore it

**Comment by EricCousineau-TRI** [comment]
Thanks! Also, seems like there're some other failures on `master` regarding the `tpu` accelerator - also seem to be caused by something outside of this PR?

**Review by justusschock — APPROVED** [review]

**Comment by EricCousineau-TRI** [comment]
Ẇewt! Thanks for y'all's patience!
